onnx/cpp2py_export.cc

```diff
@@ -115,13 +115,215 @@ PYBIND11_MODULE(onnx_cpp2py_export, onnx_cpp2py_export) {
+            return self.GetTypeStr();
+          })
+      .def_property_readonly(
+          "isHomogeneous",
+          [](const OpSchema::FormalParameter& self) {
+            auto warnings = py::module::import("warnings");
+            warnings.attr("warn")(
+                "OpSchema.FormalParameter.isHomogeneous is deprecated. Use OpSchema.FormalParameter.is_homogeneous instead.");
+            return self.GetIsHomogeneous();
+          })
+      .def_property_readonly("differentiationCategory", [](const OpSchema::FormalParameter& self) {
+        auto warnings = py::module::import("warnings");
+        warnings.attr("warn")(
+            "OpSchema.FormalParameter.differentiationCategory is deprecated. Use OpSchema.FormalParameter.differentiation_category instead.");
+        return self.GetDifferentiationCategory();
+      });
+
+  op_schema
+      .def(
+          py::init([](std::string name,
+                      std::string domain,
+                      int since_version,
+                      std::string doc,
+                      std::vector<OpSchema::FormalParameter> inputs,
+                      std::vector<OpSchema::FormalParameter> outputs,
+                      std::vector<std::tuple<std::string, std::vector<std::string>, std::string>> type_constraints,
+                      std::vector<OpSchema::Attribute> attributes) {
+            auto self = OpSchema();
+
+            self.SetName(name).SetDomain(domain).SinceVersion(since_version).SetDoc(doc);
+            // Add inputs and outputs
+            for (auto i = 0; i < inputs.size(); ++i) {
+              const OpSchema::FormalParameter& input = inputs[i];
+              self.Input(
+                  i,
+                  input.GetName(),
+                  input.GetDescription(),
+                  input.GetTypeStr(),
+                  input.GetOption(),
+                  input.GetIsHomogeneous(),
+                  input.GetMinArity(),
+                  input.GetDifferentiationCategory());
+            }
+            for (auto i = 0; i < outputs.size(); ++i) {
+              const OpSchema::FormalParameter& output = outputs[i];
+              self.Output(
+                  i,
+                  output.GetName(),
+                  output.GetDescription(),
+                  output.GetTypeStr(),
+                  output.GetOption(),
+                  output.GetIsHomogeneous(),
+                  output.GetMinArity(),
+                  output.GetDifferentiationCategory());
+            }
+            // Add type constraints
+            for (const auto& type_constraint : type_constraints) {
+              std::string type_str;
+              std::vector<std::string> constraints;
+              std::string description;
+              tie(type_str, constraints, description) = type_constraint;
+              self.TypeConstraint(type_str, constraints, description);
+            }
```

Comment:
can use structured bindings and move from `type_constraints`
```suggestion
            for (auto& [type_str, constraints, description] : type_constraints) {
              self.TypeConstraint(std::move(type_str), std::move(constraints), std::move(description));
            }
```

Reply:
This looks great! I did a quick search and it looks like structured bindings are since c++17? @jcwchen is 17 the lowest version we support?

Reply:
https://github.com/onnx/onnx/blob/515361a12ec6dace2e99f761490cb55fc2696168/CMakeLists.txt#L60-L64

No. Only Windows build requires C++17 or higher. We didn't set the corresponding required version for other platforms, because some of partners might still use lower C++ versions. Please let me know if you really want C++17 features and we can review the upgrade, although I think it might be too close to the upcoming release and can only happen to the one after that.

Reply:
SG